Objective-C/CBLDocument.mm

```diff
@@ -198,80 +201,102 @@ - (NSUInteger)countByEnumeratingWithState: (NSFastEnumerationState *)state
                                   objects: (id __unsafe_unretained [])buffer
                                     count: (NSUInteger)len
 {
-    return [_dict countByEnumeratingWithState: state objects: buffer count: len];
+    return [self.dict countByEnumeratingWithState: state objects: buffer count: len];
 }
 
 
 #pragma mark - SUBSCRIPTING
 
 
 - (CBLFragment*) objectForKeyedSubscript: (NSString*)key {
-    return [_dict objectForKeyedSubscript: key];
+    return [self.dict objectForKeyedSubscript: key];
 }
 
 
 #pragma mark - INTERNAL
 
 
+- (CBLDatabase*) database {
+    @synchronized (self.lock) {
+        return _database;
+    }
+}
+
+
 - (void) setDatabase: (CBLDatabase *)database {
-    _database = database;
-    _c4db = _database.c4db;
+    @synchronized (self.lock) {
+        _database = database;
+        _c4db = _database.c4db;
+    }
 }
 
 
 - (BOOL) isEmpty {
-    return _dict.isEmpty;
+    return self.dict.isEmpty;
 }
 
 
 - (BOOL) save: (NSError**)outError {
-    return [self saveWithConflictResolver: self.effectiveConflictResolver
-                                 deletion: NO
-                                    error: outError];
+    @synchronized (self.lock) {
+        return [self saveWithConflictResolver: self.effectiveConflictResolver
+                                     deletion: NO
+                                        error: outError];
+    }
 }
 
 
 - (BOOL) deleteDocument: (NSError**)outError {
-    return [self saveWithConflictResolver: self.effectiveConflictResolver
-                                 deletion: YES
-                                    error: outError];
+    @synchronized (self.lock) {
```

Comment:
Lock with `saveWithConflictResolver ` method is enough?? the same question applies to all method call `saveWithConflictResolver `.